plan/new_physical_plan_builder.go

```diff
@@ -98,6 +98,91 @@ func (p *Projection) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, e
 	return task, p.storeTaskProfile(prop, task)
 }
 
+// convert2NewPhysicalPlan implements PhysicalPlan interface.
+// Join has three physical operators: Hash Join, Merge Join and Index Look Up Join. We implement Hash Join at first.
+func (p *LogicalJoin) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, error) {
+	task, err := p.getTaskProfile(prop)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if task != nil {
+		return task, nil
+	}
+	switch p.JoinType {
+	case SemiJoin, LeftOuterSemiJoin:
+		task, err = p.convert2SemiJoin()
+	default:
+		// TODO: We will consider smj and index look up join in the future.
+		task, err = p.convert2HashJoin()
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Because hash join is executed by multiple goroutines, it will not propagate physical property any more.
+	// TODO: We will consider the problem of property again for parallel execution.
+	task = prop.enforceProperty(task, p.ctx, p.allocator)
+	return task, p.storeTaskProfile(prop, task)
+}
+
+func (p *LogicalJoin) convert2SemiJoin() (taskProfile, error) {
+	lChild := p.children[0].(LogicalPlan)
+	rChild := p.children[1].(LogicalPlan)
+	semiJoin := PhysicalHashSemiJoin{
+		WithAux:         LeftOuterSemiJoin == p.JoinType,
+		EqualConditions: p.EqualConditions,
+		LeftConditions:  p.LeftConditions,
+		RightConditions: p.RightConditions,
+		OtherConditions: p.OtherConditions,
+		Anti:            p.anti,
+	}.init(p.allocator, p.ctx)
+	semiJoin.SetSchema(p.schema)
+	lTask, err := lChild.convert2NewPhysicalPlan(&requiredProp{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rTask, err := rChild.convert2NewPhysicalPlan(&requiredProp{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return semiJoin.attach2TaskProfile(lTask, rTask), nil
+}
+
+func (p *LogicalJoin) convert2HashJoin() (taskProfile, error) {
+	lChild := p.children[0].(LogicalPlan)
+	rChild := p.children[1].(LogicalPlan)
+	hashJoin := PhysicalHashJoin{
+		EqualConditions: p.EqualConditions,
+		LeftConditions:  p.LeftConditions,
+		RightConditions: p.RightConditions,
+		OtherConditions: p.OtherConditions,
+		JoinType:        p.JoinType,
+		Concurrency:     JoinConcurrency,
+		DefaultValues:   p.DefaultValues,
+	}.init(p.allocator, p.ctx)
+	hashJoin.SetSchema(p.schema)
+	lTask, err := lChild.convert2NewPhysicalPlan(&requiredProp{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rTask, err := rChild.convert2NewPhysicalPlan(&requiredProp{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	switch p.JoinType {
+	case LeftOuterJoin:
+		hashJoin.SmallTable = 1
+	case RightOuterJoin:
+		hashJoin.SmallTable = 0
+	case InnerJoin:
+		// We will use right table as small table.
```

Comment:
If right plan's count is smaller than the left one, we will treat right plan as small table.